CRM/Campaign/BAO/Petition.php

```diff
@@ -102,33 +102,24 @@ public function createSignature(&$params) {
    * @return bool
    */
   public function confirmSignature($activity_id, $contact_id, $petition_id) {
-    // change activity status to completed (status_id = 2)
-    // I wonder why do we need contact_id when we have activity_id anyway? [chastell]
-    $sql = 'UPDATE civicrm_activity SET status_id = 2 WHERE id = %1';
-    $activityContacts = CRM_Activity_BAO_ActivityContact::buildOptions('record_type_id', 'validate');
-    $sourceID = CRM_Utils_Array::key('Activity Source', $activityContacts);
-    $params = [
-      1 => [$activity_id, 'Integer'],
-      2 => [$contact_id, 'Integer'],
-      3 => [$sourceID, 'Integer'],
-    ];
-    CRM_Core_DAO::executeQuery($sql, $params);
+    // change activity status to completed
+    \Civi\Api4\Activity::update(FALSE)
+      ->addValue('status_id:name', 'Completed')
+      ->addWhere('id', '=', $activity_id)
+      ->execute();
+    \Civi\Api4\ActivityContact::update(FALSE)
+      ->addValue('contact_id', $contact_id)
+      ->addWhere('activity_id', '=', $activity_id)
+      ->addWhere('record_type_id:name', '=', 'Activity Source')
+      ->execute();
 
-    $sql = 'UPDATE civicrm_activity_contact SET contact_id = %2 WHERE activity_id = %1 AND record_type_id = %3';
-    CRM_Core_DAO::executeQuery($sql, $params);
     // remove 'Unconfirmed' tag for this contact
-    $tag_name = Civi::settings()->get('tag_unconfirmed');
+    \Civi\Api4\EntityTag::delete(FALSE)
+      ->addWhere('tag_id:name', '=', Civi::settings()->get('tag_unconfirmed'))
```

Comment:
Ok, I confirmed that the api won't crash if it tries to delete a tag that doesn't exist, so this will be fine even if the tag is missing.